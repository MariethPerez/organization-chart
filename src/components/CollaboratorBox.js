import React from "react";
import styled from "@emotion/styled";

import CollaboratorForm from "./CollaboratorForm";

function CollaboratorBox({ collaborator, getChildren, addChild }) {
  const [openedForm, setOpenedForm] = React.useState(false);

  const children = getChildren(collaborator.id);

  function addNewChild(collabName) {
    addChild({ id: Date.now(), name: collabName, parentId: collaborator.id });
  }

  const Container = styled.div`
    width: 100%;
    display: flex;
    justify-content: center;
    margin-bottom: 30px;
  `;

  const Box = styled.div`
    display: flex;
    flex-direction: column;
    align-items: center;
    justify-content: center;
    width: 250px;
    height: 60px;
    padding: 1.5em;
    background: black;
    border-radius: 0.5em;
  `;

  const ColaboratorName = styled.p`
    font-size: 2em;
    color: white;
    text-decoration: none;
  `;

  const AddButton = styled.div`
    border: 2px solid darkgray;
    border-radius: 50%;
    width: 50px;
    height: 50px;
    font-size: 50px;
    display: flex;
    justify-content: center;
    align-items: center;
    background-color: gray;
    color: white;
    cursor: pointer;
    text-decoration: none;
    z-index: 1;
    &:hover {
      color: red;
      border-color: red;
    }
  `;

  const StyledUl = styled.ul`
    list-style: none;
  `;

  function handleClick() {
    setOpenedForm(true);
  }

  return (
    <>
      <Container>
        <Box>
          <ColaboratorName>{collaborator.name}</ColaboratorName>
          <AddButton onClick={handleClick}>+</AddButton>
        </Box>
      </Container>
      <StyledUl>
        {children.map(collaborator => {
          return (
            <li key={collaborator.id}>
              <CollaboratorBox
                collaborator={collaborator}
                getChildren={getChildren}
                addChild={addChild}
              />
            </li>
          );
        })}
      </StyledUl>
      {openedForm && (
        <CollaboratorForm
          addNewChild={addNewChild}
          setOpenedForm={setOpenedForm}
        />
      )}
    </>
  );
}

export default CollaboratorBox;
