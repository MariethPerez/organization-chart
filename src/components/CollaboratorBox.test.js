import React from "react";
import { render, fireEvent, waitForElement } from "@testing-library/react";

import CollaboratorBox from "./CollaboratorBox";

test("CollaboratorBox", async () => {
  const getChildren = jest.fn();
  const addChild = ["jose"];

  const { getByLabelText } = render(
    <CollaboratorBox addChild={addChild} getChildren={getChildren} />
  );

  const button = getByLabelText("Add Button");
  fireEvent.submit(button);

  const input = await waitForElement(() => getByLabelText("Collaborator Name"));

  fireEvent.change(input, {
    target: {
      value: "juancito"
    }
  });

  fireEvent.submit(input);

  expect(getChildren).toHaveBeenCalledWith("juancito");
});

test("CollaboratorBox UI", async () => {
  const { getByLabelText, asFragment} = 
  render(
    <CollaboratorBox addChild={["jose"]}
    getChildren={jest.fn()} />
  );

  expect(asFragment()).toMatchSnapshot();

  fireEvent.click(getByLabelText("Add Button"));
  await waitForElement(()=> getByLabelText("Collaborator name"));
  expect(asFragment()).toMatchSnapshot();

  fireEvent.click(getByLabelText("");
  expect(asFragment()).toMatchSnapshot();

});