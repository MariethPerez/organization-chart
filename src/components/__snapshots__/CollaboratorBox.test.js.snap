// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CollaboratorBox UI 1`] = `
<DocumentFragment>
  .emotion-3 {
  width: 100%;
  margin-bottom: 30px;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 250px;
  height: 60px;
  padding: 1.5em;
  background: black;
  border-radius: 0.5em;
  position: relative;
}

.emotion-2:before {
  content: "";
  border-bottom: solid 3px black;
  border-left: solid 3px black;
  position: absolute;
  top: -30px;
  width: 49px;
  height: 80px;
  left: -49px;
}

.emotion-0 {
  font-family: sans-serif;
  font-size: 2em;
  color: white;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.emotion-1 {
  border: 2px solid darkgray;
  border-radius: 50%;
  width: 50px;
  height: 50px;
  font-size: 50px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: gray;
  color: white;
  -webkit-text-decoration: none;
  text-decoration: none;
  z-index: 1;
  cursor: pointer;
}

.emotion-1:hover {
  background-color: #484848;
}

<div
    class="emotion-3"
  >
    <div
      class="emotion-2"
    >
      <p
        class="emotion-0"
      />
      <div
        aria-label="Add Button"
        class="emotion-1"
      >
        +
      </div>
    </div>
  </div>
  .emotion-0 {
  list-style: none;
  padding-left: 120px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

<ul
    class="emotion-0"
  />
</DocumentFragment>
`;

exports[`CollaboratorBox UI 2`] = `
<DocumentFragment>
  .emotion-3 {
  width: 100%;
  margin-bottom: 30px;
}

.emotion-2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 250px;
  height: 60px;
  padding: 1.5em;
  background: black;
  border-radius: 0.5em;
  position: relative;
}

.emotion-2:before {
  content: "";
  border-bottom: solid 3px black;
  border-left: solid 3px black;
  position: absolute;
  top: -30px;
  width: 49px;
  height: 80px;
  left: -49px;
}

.emotion-0 {
  font-family: sans-serif;
  font-size: 2em;
  color: white;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.emotion-1 {
  border: 2px solid darkgray;
  border-radius: 50%;
  width: 50px;
  height: 50px;
  font-size: 50px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: gray;
  color: white;
  -webkit-text-decoration: none;
  text-decoration: none;
  z-index: 1;
  cursor: pointer;
}

.emotion-1:hover {
  background-color: #484848;
}

<div
    class="emotion-3"
  >
    <div
      class="emotion-2"
    >
      <p
        class="emotion-0"
      />
      <div
        aria-label="Add Button"
        class="emotion-1"
      >
        +
      </div>
    </div>
  </div>
  .emotion-0 {
  list-style: none;
  padding-left: 120px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

<ul
    class="emotion-0"
  />
  .emotion-5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  position: fixed;
  z-index: 2;
  left: 0;
  top: 0;
  width: 100%;
  height: 100%;
  overflow: auto;
  background-color: rgba(0,0,0,0.4);
}

.emotion-4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 500px;
  padding: 1.5em;
  background: black;
  border-radius: 0.5em;
  margin: 0 auto;
}

.emotion-0 {
  font-size: 2em;
  cursor: pointer;
  color: white;
  text-align: right;
  width: 100%;
}

.emotion-1 {
  text-align: center;
  margin: 15px;
  font-size: 40px;
  color: white;
  font-weight: bolder;
}

.emotion-2 {
  margin: 1em 0;
  font-size: 30px;
  padding-top: 10px;
  padding-bottom: 10px;
  border-radius: 4px;
  outline: none;
  border: none;
  border-bottom: 1px solid #ddd;
  width: 100%;
  text-align: center;
}

.emotion-2:focus {
  border: 1px solid red;
}

.emotion-3 {
  line-height: 1.5em;
  background: darkgray;
  border: none;
  color: white;
  font-size: 20px;
  font-weight: bold;
  padding: 0.5em;
  border-radius: 0.25rem;
  text-align: center;
  width: 50%;
  cursor: pointer;
}

.emotion-3:hover {
  background-color: #484848;
}

<div
    class="emotion-5"
  >
    <form
      class="emotion-4"
    >
      <span
        class="emotion-0"
      >
        Ã—
      </span>
      <h1
        class="emotion-1"
      >
        Collaborator name
      </h1>
      <label
        aria-label="Write Your Name"
        name="name"
      >
        <input
          aria-label="Get Collaborator Name"
          class="emotion-2"
          id="my_name"
          name="name"
          placeholder="Enter colaborator name"
          type="text"
        />
      </label>
      <input
        class="emotion-3"
        type="submit"
        value="submit"
      />
    </form>
  </div>
</DocumentFragment>
`;
